core/kernel/user_access.c

```diff
@@ -74,6 +96,137 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
 	return res;
 }
 
+void *bb_alloc(size_t len)
+{
+	struct user_mode_ctx *uctx = get_current_uctx();
+	size_t offs = 0;
+	void *bb = NULL;
+
+	if (uctx && !ADD_OVERFLOW(uctx->bbuf_offs, len, &offs) &&
+	    offs <= uctx->bbuf_size) {
+		bb = uctx->bbuf + uctx->bbuf_offs;
+		uctx->bbuf_offs = ROUNDUP(offs, BB_ALIGNMENT);
+	}
+	return bb;
+}
+
+static void bb_free_helper(struct user_mode_ctx *uctx, vaddr_t bb, size_t len)
+{
+	vaddr_t bbuf = (vaddr_t)uctx->bbuf;
+
+	if (bb > bbuf && IS_ALIGNED(bb, BB_ALIGNMENT)) {
```

Comment:
The first bounce buffer will be allocated at `bbuf`. We are missing to free them as we check `bb > bbuf`, here.

Reply:
Good catch. We should use `>=` instead.

Reply:
Are you going to fix this on `jenswi-linaro/bounce_buffer`, or shall I fix that in a new PR?